aidial_assistant/chain/command_chain.py

```diff
@@ -86,27 +108,39 @@ async def run_chat(self, history: History, callback: ChainCallback):
                 else history.to_client_messages()
             )
             await self._generate_result(messages, callback)
-        except InvalidRequestError as e:
-            if dialogue.is_empty() or e.code == "429":
+        except (InvalidRequestError, LimitExceededException) as e:
+            if dialogue.is_empty() or (
+                isinstance(e, InvalidRequestError) and e.code == "429"
+            ):
                 raise
 
             # Assuming the context length is exceeded
             dialogue.pop()
+            # TODO: Limit the error message size. The error message should not exceed reserved assistant overheads.
             await self._generate_result(
                 history.to_best_effort_messages(str(e), dialogue), callback
             )
 
     async def _run_with_protocol_failure_retries(
-        self, callback: ChainCallback, messages: list[Message]
-    ) -> Tuple[str, str] | None:
+        self,
+        callback: ChainCallback,
+        messages: list[Message],
+        model_request_limiter: ModelRequestLimiter | None = None,
+    ) -> Exchange | None:
         last_error: Exception | None = None
         try:
             self._log_messages(messages)
             retries = Dialogue()
             while True:
+                all_messages = self._reinforce_json_format(
+                    messages + retries.messages
+                )
+                if model_request_limiter:
+                    await model_request_limiter.verify_limit(all_messages)
+
                 chunk_stream = CumulativeStream(
                     self.model_client.agenerate(
-                        self._reinforce_json_format(messages + retries.messages)
+                        all_messages, **self.model_extra_args  # type: ignore
```

Comment:
`max_tokens` must be decreased each time we add a new turn to the dialogue by the token size of this turn, right?

Reply:
This parameter limits completions in nested (addon) sessions. It is used for generating commands and replies within these sessions.